astroid/inference.py

```diff
@@ -1063,20 +1077,26 @@ def _cached_generator(func, instance, args, kwargs, _cache={}):  # noqa: B006
 # are mutated with a new instance of the property. This is why we cache the result
 # of the function's inference.
 @_cached_generator
-def infer_functiondef(self, context=None):
+def infer_functiondef(
+    self: nodes.FunctionDef, context: Optional[InferenceContext] = None
+) -> Generator[
+    Union["astroid_objects.Property", nodes.FunctionDef],
+    None,
+    Dict[str, Union[nodes.FunctionDef, InferenceContext, None]],
```

Comment:
A `Union` value type for dicts is not really helpful. We should either use `Dict[str, Any]` or a `TypedDict`.

Reply:
Doesn't this narrow down the type from `Any`?  It's the results of these return values:

https://github.com/PyCQA/astroid/blob/393b27660f63c5c93d6e3bdb23da3534065564b2/astroid/inference.py#L1069
https://github.com/PyCQA/astroid/blob/393b27660f63c5c93d6e3bdb23da3534065564b2/astroid/inference.py#L1081

How would you type those?

Reply:
The type itself is correct, it's just not very useful.

Think of it as a consumer of this function. What would you do if you know the expected return type is `Dict[str, Union[...]]`? You would need to combine every attribute access to say `d.node` with an isinstance call to narrow down the type. That is just impractical. If it's instead typed as `Any`, mypy will just skip it.

The alternative would be a TypeDict. Maybe that's even more useful since other inference functions seem to return the same dict. Something like this, also I'm not sure about the name itself. Would need to check if it's actual an `InferenceError`.
```py
class InferenceError(TypedDict):
    node: nodes.NodeNG
    context: Optional[InferenceContext]
```

Reply:
A little breakpointing showed that it is used here:
https://github.com/PyCQA/astroid/blob/393b27660f63c5c93d6e3bdb23da3534065564b2/astroid/inference.py#L1047-L1057
(Might be used elsewhere of course).

So here the return value of `infer_functiondef` is part of the key in cache.
Perhaps `class NodeContextPair`?

Reply:
Let's use that for now. I'll come back to it once I'm done reviewing the first doc PR.